### PR TITLE
dont set model scale or mu in GLM

### LINKED ISSUE #4016: GLM attributes
Just based on a "this looks funny" [knee-jerk:](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/genmod/generalized_linear_model.py#L1203)

```
        result = fit_elasticnet(self, method=method,
                                alpha=alpha,
                                start_params=start_params,
                                refit=refit,
                                **defaults)

        self.mu = self.predict(result.params)
        self.scale = self.estimate_scale(self.mu)

        return result
```

Is this supposed to be setting `self.mu` and `self.scale`?  It seems more appropriate that it would set `result.mu` and `result.scale`.

### PR DESCRIPTION
**Pull request by jbrockmendel**
Closes #4016

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Does coveralls have a config option to ignore decreases of a certain size?  The "failure" here is a false-positive.

**Comment by josef-pkt** [comment]
I think this can wait for the other refactorings, and after glm weights.

my guess is that this PR is fine
see also #2399 and #3357

**Comment by josef-pkt** [comment]
failing unit tests in several GLM cases across all environments

**Comment by jbrockmendel** [comment]
Yep, I'll take a look.  In the meantime #5147 is there to help.

File: statsmodels/genmod/generalized_linear_model.py [hunk 1]
```diff
@@ -791,7 +793,7 @@ def estimate_scale(self, mu):
                 return self._estimate_x2_scale(mu)
             elif self.scaletype.lower() == 'dev':
                 return (self.family.deviance(self.endog, mu, self.var_weights,
-                                             self.freq_weights, self.scale) /
+                                             self.freq_weights, scale) /
```
Comment (josef-pkt) [inline statsmodels/genmod/generalized_linear_model.py]: Only based on spot checking
my guess is that scale here should be 1.
scale should not be an argument in estimate_scale

**Comment by josef-pkt** [comment]
when you test locally, you should include the failing test to see whether it's fixed

`pytest statsmodels.base.tests.test_generic_methods::TestGenericGLM`
or with full path to test_generic_methods

**Comment by jbrockmendel** [comment]
Hmm doesn't seem to have fixed it

**Comment by josef-pkt** [comment]
Ok, this needs a look at `TestGenericGLM.test_zero_constrained` and `_fit_zeros`

The fail is in `bse` and it looks like the scale differs, maybe it uses the wrong df_resid

```
>>> x = np.array([0.064275, 0.065696, 0.064033])
>>> y = np.array([0.064219, 0.065638, 0.063977])
>>> x / y
array([ 1.00087202,  1.00088363,  1.00087531])
>>> (x / y)**2
array([ 1.00174479,  1.00176805,  1.0017514 ])
```

**Comment by josef-pkt** [comment]
without investigating an example:
I guess the problem is when `results.df_resid != model.df_resid` which is the case in fit_constrained and similar.
`GLM.estimate_scale` uses `model.df_resid`

possibly fit_zeros needs more unit tests for GLM when scale!=1, currently the failing TestGenericGLM is the only one
#4195 fixed scale for fit_constrained

**Comment by josef-pkt** [comment]
I haven't tried to figure out why scale in master is correct, i.e. unit test passes in master for fit_zeros.

ok, base.model `_fit_zeros` has special casing to account for current glm `model.scale`

```
        # TODO: do we need to change res._results.scale in some models?
        if hasattr(res_constr.model, 'scale'):
            # Note: res.model is self
            # GLM problem, see #2399,
            # TODO: remove from model if not needed anymore
            res.model.scale = res._results.scale = res_constr.model.scale
```

my guess is the last line needs to be change to something like `res._results.scale = res_constr.scale`

**Comment by josef-pkt** [comment]
maybe another redesign that keeps the scale attribute in model
#4664

**Comment by josef-pkt** [comment]
This PR cannot be merged without further thought and decisions. We keep it open to see one possible version.

We need temporary model attributes for the implementation of `fit`, but we don't want those attributes to cause stale state for the results instance.
I'm still not sure how to resolve this dual role. see new issues for auxiliary parameters.
(GAM alpha or pen_weight is even worse to support in a clean way as fit argument)

**Comment by jbrockmendel** [comment]
@josef-pkt IIRC you had a (more recent) PR that addressed a related issue.  Did you get anywhere with this?

**Comment by bashtage** [comment]
@jbrockmendel Could you rebase so this can be considered?

**Pull request closed without merging**